daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/NullInfosetInputter.scala

```diff
@@ -0,0 +1,123 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.infoset
+
+import java.io.InputStream
+
+import scala.collection.mutable.ArrayBuffer
+
+import scala.xml.Elem
+import scala.xml.Node
+import scala.xml.SAXParser
+import scala.xml.Text
+import scala.xml.XML
+
+import org.apache.daffodil.dpath.NodeInfo
+import org.apache.daffodil.infoset.InfosetInputterEventType._
+import org.apache.daffodil.util.MaybeBoolean
+import org.apache.daffodil.xml.DaffodilSAXParserFactory
+import org.apache.daffodil.xml.XMLUtils
+
+object NullInfosetInputter {
+
+  case class Event(
+    eventType: InfosetInputterEventType,
+    localName: String = null,
+    namespaceURI: String = null,
+    simpleText: String = null,
+    isNilled: MaybeBoolean = MaybeBoolean.Nope,
+  )
+
+  def createEvents(is: InputStream): Array[Event] = {
+    val elem = {
+      val parser: SAXParser = {
+        val f = DaffodilSAXParserFactory()
+        f.setNamespaceAware(false)
+        val p = f.newSAXParser()
+        p
+      }
+      val node = XML.withSAXParser(parser).load(is)
+      val normalized = XMLUtils.normalize(node)
+      normalized
+    }
+    val events = ArrayBuffer[Event]()
+    events += Event(StartDocument)
+    nodeToEvents(elem, events)
+    events += Event(EndDocument)
+    events.toArray
+  }
+
+  /**
+   * Recursively walk a Scala XML Node, converting each Elem to an Event
+   * and appends it to the events array
+   */
+  private def nodeToEvents(node: Node, events: ArrayBuffer[Event]): Unit = {
+    node match {
+      case elem: Elem => {
+        val isSimple = elem.child.length == 0 || elem.child(0).isInstanceOf[Text]
+        val localName = elem.label
+        val namespaceURI = elem.namespace
+        val (simpleText, isNilled) = if (isSimple) {
+          val text = XMLUtils.remapPUAToXMLIllegalCharacters(elem.text)
+          val isNilled = elem.attribute(XMLUtils.XSI_NAMESPACE, "nil").map { attrs =>
+            val str = attrs.head.toString
+            val value = str == "true" || str == "1"
+            MaybeBoolean(value)
+          }.getOrElse(MaybeBoolean.Nope)
+          (text, isNilled)
+        } else {
+          (null, MaybeBoolean.Nope)
+        }
+
+        events += Event(StartElement, localName, namespaceURI, simpleText, isNilled)
+        if (!isSimple) elem.child.foreach { nodeToEvents(_, events) }
+        events += Event(EndElement, localName, namespaceURI)
+      }
+      case _ => // ignore
```

Comment:
I think really this function should just expect an `Elem` instead of `Node`. Due to the `XMLutils.normalize` call it's impossible for the node parameter to not be a `Elem`. Will update it.